bootstrap/kubeadm/api/v1alpha3/kubeadmconfig_webhook.go

```diff
@@ -17,11 +17,91 @@ limitations under the License.
 package v1alpha3
 
 import (
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
 	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+var (
+	ConflictingFileSourceMsg = "only one of content of contentFrom may be specified for a single file"
+	MissingFileSourceMsg     = "source for file content must be specified if contenFrom is non-nil"
+	MissingSecretNameMsg     = "secret file source must specify non-empty secret name"
+	MissingSecretKeyMsg      = "secret file source must specify non-empty secret key"
 )
 
-func (r *KubeadmConfig) SetupWebhookWithManager(mgr ctrl.Manager) error {
+func (c *KubeadmConfig) SetupWebhookWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewWebhookManagedBy(mgr).
-		For(r).
+		For(c).
 		Complete()
 }
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-bootstrap-cluster-x-k8s-io-v1alpha3-kubeadmconfig,mutating=false,failurePolicy=fail,matchPolicy=Equivalent,groups=bootstrap.cluster.x-k8s.io,resources=kubeadmconfigs,versions=v1alpha3,name=validation.kubeadmconfig.bootstrap.cluster.x-k8s.io,sideEffects=None
+
+var _ webhook.Validator = &KubeadmConfig{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (c *KubeadmConfig) ValidateCreate() error {
+	return c.validate()
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (c *KubeadmConfig) ValidateUpdate(old runtime.Object) error {
+	return c.validate()
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (c *KubeadmConfig) ValidateDelete() error {
+	return nil
+}
+
+func (c *KubeadmConfig) validate() error {
```

Comment:
```suggestion
func (c *KubeadmConfigSpec) validate() error {
```

If we add this method to `KubeadmConfigSpec` instead of `KubeadmConfig` we should be able to call it from the webhooks defined under `controlplane/kubeadm` as well.

Reply:
better?